src/hato/client.clj

```diff
@@ -284,42 +275,114 @@
                       (->BodyPublisher req))
              .build)))
 
+(defn unwrap-async-exception
+  "Helper function to unwrap common async exceptions to get the root cause such as
+  an IOException."
+  [^Exception ex]
+  (if (or (instance? CompletionException ex)
+          (instance? ExecutionException ex))
+    (if-let [cause (.getCause ex)]
+      (unwrap-async-exception cause)
+      ex)
+    ex))
+
+(def retry-exceptions
```

Comment:
Do you think these should be in client.clj or should we move to a separate namespace?